tf2onnx/graph.py

```diff
@@ -1207,43 +1248,92 @@ def insert_new_node_on_output(self, op_type, output_name, name, domain=None, **k
         new_output = port_name(name)
         new_node = self.make_node(op_type, [output_name], attr=kwargs, outputs=[new_output], name=name, domain=domain)
 
-        to_replace = [n for n in self.get_nodes() if n != new_node]
+        to_replace = [self.get_node_by_name(n) for n in self._input_to_node_name[output_name]]
+        to_replace = [n for n in to_replace if n != new_node]
         self.replace_all_inputs(to_replace, output_name, new_output)
         return new_node
 
     def find_output_consumers(self, output_name):
         """Find all nodes consuming a given output."""
+        if output_name in self._input_to_node_name:
+            ops = self._input_to_node_name[output_name]
+            ops = [self.get_node_by_name(n) for n in ops]
+        else:
+            ops = []  # self.get_nodes()
         nodes = []
-        for node in self.get_nodes():
+        for node in ops:
+            if node is None:
+                continue
             if output_name in node.input:
                 nodes.append(node)
 
-            # find consumers in sub graphs
-            body_graphs = node.get_body_graphs()
-            if body_graphs:
-                for g in body_graphs.values():
-                    nodes.extend(g.find_output_consumers(output_name))
+        # find consumers in sub graphs
+        if output_name in self._input_to_graph:
+            for g in self._input_to_graph[output_name].values():
+                nodes.extend(g.find_output_consumers(output_name))
         return nodes
 
-    @staticmethod
-    def replace_all_inputs(ops, old_input, new_input):
-        """Replace all inputs pointing to old_input with new_input."""
+    def _register_input_name(self, input_name, node, only_graph=False):
+        "Register node taking a specific input."
+        if not only_graph:
+            if input_name not in self._input_to_node_name:
+                self._input_to_node_name[input_name] = set()
+            self._input_to_node_name[input_name].add(node.name)
+        if self.parent_graph is not None:
+            if input_name not in self.parent_graph._input_to_graph:
+                self.parent_graph._input_to_graph[input_name] = {}
+            self.parent_graph._input_to_graph[input_name][id(self)] = self
+            self.parent_graph._register_input_name(input_name, node, only_graph=True)
+
+    def _unregister_input_name(self, input_name, node, only_graph=False):
+        "Unregister node taking a specific input."
+        node_name = node.name
+        if not only_graph:
+            if input_name in self._input_to_node_name[input_name]:
```

Comment:
```suggestion
            if input_name in self._input_to_node_name:
```

Reply:
Good catch. I'm worrried that the unit test pass then. This line was probably never used.

Reply:
Done

Reply:
I think this passes because there is no test to make sure the index doesn't contain extraneous entries.  As I think you mentioned before, if the index has extra entries it isn't actually an issue as far as correctness is concerned and it only causes a slight performance decrease.  Still, we may want to consider adding a validate_indices function that runs during unit tests between, say, optimization passes and once after the rewriters/handlers finish.  What do you think?

Reply:
I'm mostly concerned that later someone will forget to update the indices and introduce a subtle bug that is hard to catch.

Reply:
I don't mind going further but I'll have to look at every call to replace_all_inputs and possibly change them. This is a huge refactoring. The PR will grow. One big PR or two smaller ones but not that small either, that is the question.

Reply:
maybe we can have a check_graph() method on the graph class that we can call from some places (ie. in ut there would be a good place to call this between conversion and optimizer) that validates all is in order?

Reply:
Same PR or another one?